trading_calendars/exchange_calendar_xhkg.py

```diff
@@ -0,0 +1,271 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import time
+from itertools import chain
+
+import numpy as np
+import pandas as pd
+from pandas.tseries.holiday import (
+    EasterMonday,
+    GoodFriday,
+    Holiday,
+    sunday_to_monday,
+)
+from pytz import timezone
+
+from .trading_calendar import (
+    TradingCalendar,
+    HolidayCalendar,
+    MONDAY,
+    TUESDAY,
+    WEDNESDAY,
+    THURSDAY,
+    FRIDAY,
+    SATURDAY,
+    SUNDAY,
+)
+from .common_holidays import (
+    boxing_day,
+    chinese_buddhas_birthday_dates,
+    chinese_lunar_new_year_dates,
+    christmas,
+    day_after_mid_autumn_festival_dates,
+    double_ninth_festival_dates,
+    dragon_boat_festival_dates,
+    national_day,
+    new_years_day,
+    qingming_festival_dates,
+    weekend_christmas,
+)
+from .utils.pandas_utils import vectorized_sunday_to_monday
+
+# Useful resources for making changes to this file:
+# # /etc/lunisolar
+# http://www.math.nus.edu.sg/aslaksen/calendar/cal.pdf
+# https://www.hko.gov.hk/gts/time/calendarinfo.htm
+#   - the almanacs on this page are also useful
+
+
+labor_day = Holiday(
+    'Labor Day',
+    month=5,
+    day=1,
+    observance=sunday_to_monday,
+)
+
+establishment_day = Holiday(
+    'Hong Kong Special Administrative Region Establishment Day',
+    month=7,
+    day=1,
+    observance=sunday_to_monday,
+)
+
+christmas_eve = Holiday(
+    'Christmas Eve',
+    month=12,
+    day=24,
+    days_of_week=(MONDAY, TUESDAY, WEDNESDAY, THURSDAY, FRIDAY),
+)
+
+
+def boxing_day_obs(dt):
+    if dt.weekday in (MONDAY, TUESDAY):
+        return dt + pd.Timedelta(days=1)
+    return dt
+
+
+class XHKGExchangeCalendar(TradingCalendar):
+    """
+    Exchange calendar for the Hong Kong Stock Exchange (XNYS).
+
+    Open Time: 9:31 AM, Asia/Hong_Kong
+    Close Time: 4:00 PM, Asia/Hong_Kong
+
+    Regularly-Observed Holidays:
+    - New Years Day (observed on monday when Jan 1 is a Sunday)
+    - Lunar New Year and the following 2 days. If the 3rd day of the lunar year
+      is a Sunday, then the next Monday is a holiday.
+    - Ching Ming Festival
+    - Good Friday
+    - Easter Monday
+    - Buddhas Birthday
+    - Dragon Boat Festival
+    - Chinese National Day (observed on monday when Oct 1 is a Sunday)
+    - Day Following Mid-Autumn Festival
+    - Chung Yeung Festival
+    - Christmas (observed on nearest weekday to December 25)
+    - Day after Christmas is observed
+
+    Regularly-Observed Early Closes:
+    - Lunar New Year's Eve
+    - Christmas Eve
+    - New Year's Eve
+
+
+    Additional Irregularities:
+    - Closes frequently for severe weather.
+    """
+    name = 'XHKG'
+    tz = timezone('Asia/Hong_Kong')
+    open_time = time(9, 31)
+    close_time = time(16)
```

Comment:
ugh, @ssanderson any ideas?

Reply:
Given that we're currently only looking to support daily data for this calendar in the near term, I think this is fine to leave for now. I'd rather do a separate pass to add support for point-in-time standard opens/closes.

Long term, we probably need to change the existing `open_time` and `close_time` properties to make them return something like a list of date ranges and standard open/close times within those ranges.

Can we open an issue here to track the fact that we don't yet have support for variable open/close times? Adding one for lunch breaks probably makes sense as well.

Reply:
FWIW, I did a quick scan downstream and I don't think anything in zipline or in any Q internal projects uses `open_time` or `close_time` directly, so I think changing this will be pretty safe.

Reply:
Honestly, given ^, I think all we need is to change the way we compute baseline opens and closes to concat a sequence of `days_at_time` calls instead of doing a single one, so maybe it's worth just doing that now?

Reply:
(The lunch break change is definitely more work than I think we want to bite off right now though).